dipy/nn/histo_resdnn.py

```diff
@@ -0,0 +1,245 @@
+                             .format(self.sh_size))
+
+        return self.model.predict(x_test)
+
+    def fit(self, data, gtab, mask=None):
+        """ Wrapper function to faciliate prediction of larger dataset.
+        The function will mask, normalize, split, predict and 're-assemble'
+        the data as a volume.
+
+        Parameters
+        ----------
+        data : np.ndarray
+            DWI signal in a 4D array
+        gtab : GradientTable class instance
+            The acquisition scheme matching the data (must contain at least
+            one b0)
+        mask : np.ndarray (optional)
+            Binary mask of the brain to avoid unnecessary computation and
+            unreliable prediction outside the brain.
+
+        Returns
+        -------
+        pred_sh_coef : np.ndarray (x, y, z, M)
+            Predicted fODF (as SH). The volume has matching shape to the input
+            data, but with ``(sh_order + 1) * (sh_order + 2) / 2`` as a last
+            dimension.
+
+        """
+        if mask is None:
+            logger.warning('Mask should be provided to accelerate '
+                           'computation, and because predictions are '
+                           'not reliable outside of the brain.')
+            mask = np.sum(data, axis=-1)
+        mask = mask.astype(bool)
+
+        # Extract B0's and obtain a mean B0
+        b0_indices = np.where(gtab.bvals < 1)[0]
+        if not len(b0_indices) > 0:
+            raise ValueError('b0 must be present for DWI normalization.')
+        logger.info('b0 indices found are: {}'.format(b0_indices))
+
+        mean_b0 = data[..., b0_indices]
+        mean_b0 = np.mean(mean_b0, axis=-1)
+
+        # Detect number of b-values and extract a single shell of DW-MRI Data
+        unique_shells = np.sort(np.unique(gtab.bvals))
+        logger.info('Number of b-values: {}'.format(unique_shells))
+
+        # Extract DWI only
+        dw_indices = np.where(gtab.bvals == unique_shells[1])[0]
+        dw_data = data[..., dw_indices]
+        dw_bvecs = gtab.bvecs[dw_indices, :]
+
+        # Normalize the DW-MRI Data with the mean b0 (voxel-wise)
+        norm_dw_data = np.zeros(dw_data.shape)
+        for n in range(len(dw_indices)):
+            norm_dw_data[..., n] = np.divide(dw_data[..., n], mean_b0,
+                                             where=np.abs(mean_b0) > 0.000001)
+
+        # Fit SH to the raw DWI signal
+        h_sphere = HemiSphere(xyz=dw_bvecs)
+        dw_sh_coef = sf_to_sh(norm_dw_data, h_sphere, smooth=0.0006,
+                              basis_type=self.basis_type,
+                              sh_order=self.sh_order)
+
+        # Flatten and mask the data (N, SH_SIZE) to facilitate chunks
+        ori_shape = dw_sh_coef.shape
+        flat_dw_sh_coef = dw_sh_coef[mask > 0]
+        flat_pred_sh_coef = np.zeros(flat_dw_sh_coef.shape)
+
+        count = len(flat_dw_sh_coef) // 1000
```

Comment:
Do you want to make 1000 something users can control? We have something like that in DTI.